packages/now-build-utils/src/detect-builders.ts

```diff
@@ -392,119 +556,409 @@ function validateFunctions(files: string[], { functions = {} }: Options) {
+      const prefix = isIndex ? '\\/' : '';
+
+      const names = [
+        isIndex ? prefix : `${fileName}\\/`,
+        prefix + escapeName(fileName),
+        featHandleMiss && cleanUrls
+          ? ''
+          : prefix + escapeName(fileName) + escapeName(ext),
+      ].filter(Boolean);
+
+      // Either filename with extension, filename without extension
+      // or nothing when the filename is `index`.
+      // When `cleanUrls: true` then do *not* add the filename with extension.
+      return `(${names.join('|')})${isIndex ? '?' : ''}`;
+    }
+
+    return segment;
+  });
+
+  const { name: fileName, ext } = parsePath(filePath);
+  const isIndex = fileName === 'index';
+  const queryString = `${query.length ? '?' : ''}${query.join('&')}`;
+
+  const src = isIndex
+    ? `^/${srcParts.slice(0, -1).join('/')}${srcParts.slice(-1)[0]}$`
+    : `^/${srcParts.join('/')}$`;
+
+  let route: Source;
+
+  if (featHandleMiss) {
+    const extensionless = ext ? filePath.slice(0, -ext.length) : filePath;
+    route = {
+      src,
+      dest: `/${extensionless}${queryString}`,
+      check: true,
+    };
+  } else {
+    route = {
+      src,
+      dest: `/${filePath}${queryString}`,
     };
   }
 
-  if (frontendBuilder) {
-    builders.push(frontendBuilder);
+  return { route, isDynamic };
+}
+
+function mergeRoutes(
+  preDefaultRoutes: Source[],
+  preDynamicRoutes: Source[],
+  outputDirectory: string,
+  apiBuilders: Builder[],
+  frontendBuilder: Builder | null,
+  options: Options
+): {
+  defaultRoutes: Route[];
+  redirectRoutes: Route[];
+} {
+  const defaultRoutes: Route[] = [];
+  const redirectRoutes: Route[] = [];
+
+  if (preDefaultRoutes && preDefaultRoutes.length > 0) {
+    if (options.featHandleMiss) {
+      defaultRoutes.push({ handle: 'miss' });
+
+      const extSet = detectApiExtensions(apiBuilders);
+
+      if (extSet.size > 0) {
+        const exts = Array.from(extSet)
+          .map(ext => ext.slice(1))
+          .join('|');
+
+        const extGroup = `(?:\\.(?:${exts}))`;
+
+        if (options.cleanUrls) {
+          redirectRoutes.push({
+            src: `^/(api(?:.+)?)/index${extGroup}?/?$`,
+            headers: { Location: options.trailingSlash ? '/$1/' : '/$1' },
+            status: 308,
+          });
+
+          redirectRoutes.push({
+            src: `^/api/(.+)${extGroup}/?$`,
+            headers: {
+              Location: options.trailingSlash ? '/api/$1/' : '/api/$1',
+            },
+            status: 308,
+          });
+        } else {
+          defaultRoutes.push({
+            src: `^/api/(.+)${extGroup}$`,
+            dest: '/api/$1',
+            check: true,
+          });
+        }
+      }
 
-    const conflictError = checkConflictingFiles(files, builders);
+      if (preDynamicRoutes) {
+        defaultRoutes.push(...preDynamicRoutes);
+      }
 
-    if (conflictError) {
-      warnings.push(conflictError);
+      if (preDefaultRoutes.length) {
```

Comment:
I'm not sure if this is correct. Is `preDefaultRoutes` only including `/api/` routes?

Reply:
@styfle yes, we don't generate routes for anything else, except for the output directory, but that happens later and they are not pushed to `preDefaultRoutes`